components/tidb_query/src/codec/mysql/time/datetime.rs

```diff
@@ -0,0 +1,1243 @@
+    #[inline]
+    get_minute, set_minute: 35, 30;
+    #[inline]
+    get_second,set_second: 29, 24;
+    #[inline]
+    get_micro, set_micro: 23, 4;
+
+    // `fsp_tt` format:
+    // | fsp: 3 bits | type: 1 bit |
+    // When `fsp` is valid (in range [0, 6]):
+    // 1. `type` bit 0 represent `DateTime`
+    // 2. `type` bit 1 represent `TimeStamp`
+    //
+    // Since `Date` does not require `fsp`, we could use `fsp` == 0b111 to represent it.
+    #[inline]
+    u8, get_fsp_tt, set_fsp_tt: 3, 0;
+}
+
+#[derive(Debug)]
+struct DateTimeConfig {
+    strict_mode: bool,
+    no_zero_in_date: bool,
+    no_zero_date: bool,
+    allow_invalid_date: bool,
+    ignore_truncate: bool,
+    time_zone: Option<Tz>,
+}
+
+impl DateTimeConfig {
+    pub fn from_ctx(ctx: &EvalContext) -> DateTimeConfig {
+        let sql_mode = ctx.cfg.sql_mode;
+        let flags = ctx.cfg.flag;
+        DateTimeConfig {
+            strict_mode: sql_mode.contains(SqlMode::STRICT_ALL_TABLES),
+            no_zero_in_date: sql_mode.contains(SqlMode::NO_ZERO_IN_DATE),
+            no_zero_date: sql_mode.contains(SqlMode::NO_ZERO_DATE),
+            allow_invalid_date: sql_mode.contains(SqlMode::INVALID_DATES),
+            ignore_truncate: flags.contains(Flag::IGNORE_TRUNCATE),
+            time_zone: Some(ctx.cfg.tz.clone()),
+        }
+    }
+}
+
+impl From<DateTimeConfig> for EvalContext {
+    fn from(config: DateTimeConfig) -> EvalContext {
+        let mut eval_config = EvalConfig::new();
+        let mut sql_mode = SqlMode::empty();
+        let mut flags = Flag::empty();
+        eval_config.tz = config.time_zone.unwrap_or_else(Tz::utc);
+
+        if config.strict_mode {
+            sql_mode |= SqlMode::STRICT_ALL_TABLES;
+        }
+        if config.allow_invalid_date {
+            sql_mode |= SqlMode::INVALID_DATES;
+        }
+        if config.no_zero_date {
+            sql_mode |= SqlMode::NO_ZERO_DATE;
+        }
+        if config.no_zero_in_date {
+            sql_mode |= SqlMode::NO_ZERO_IN_DATE;
+        }
+
+        if config.ignore_truncate {
+            flags |= Flag::IGNORE_TRUNCATE;
+        }
+
+        eval_config.set_sql_mode(sql_mode).set_flag(flags);
+
+        EvalContext::new(Arc::new(eval_config))
+    }
+}
+
+#[derive(PartialEq, Clone, Copy, Debug)]
+pub enum TimeType {
+    Date,
+    DateTime,
+    TimeStamp,
+}
+
+impl TryFrom<FieldTypeTp> for TimeType {
+    type Error = codec::Error;
+    fn try_from(time_type: FieldTypeTp) -> Result<TimeType> {
+        Ok(match time_type {
+            FieldTypeTp::Date => TimeType::Date,
+            FieldTypeTp::DateTime => TimeType::DateTime,
+            FieldTypeTp::Timestamp => TimeType::TimeStamp,
+            _ => return Err(box_err!("Time does not support field type {}", time_type)),
+        })
+    }
+}
+
+// The common set of methods for `date/time`
+impl Time {
+    /// Returns the hour number from 0 to 23.
+    #[inline]
+    pub fn hour(self) -> u32 {
+        self.get_hour()
```

Comment:
why not directly mark bitfield to use `hour` as the getter?

Reply:
if we consider `hour` as a member of bitfield, it should be public. However, the `pub` keyword will affect the setter making `set_hour` a public function